components/raftstore/src/store/util.rs

```diff
@@ -1123,6 +1125,41 @@ pub fn check_conf_change(
     }
 }
 
+pub fn check_remove_node(
+    cfg: &Config,
+    change_peers: &[ChangePeerRequest],
+    peer_heartbeat: &collections::HashMap<u64, std::time::Instant>,
+) -> Result<()> {
+    // max heartbeats missed to qualify being a slow peer
+    const MAX_MISSED_HEARTBEATS: u32 = 8;
+    let mut slow_peer_count: u32 = 0;
+    for last_heartbeat in peer_heartbeat.values() {
+        if last_heartbeat.elapsed() > MAX_MISSED_HEARTBEATS * cfg.raft_heartbeat_interval() {
+            slow_peer_count += 1;
+        }
+    }
+
+    for cp in change_peers {
+        let (change_type, peer) = (cp.get_change_type(), cp.get_peer());
+        if change_type == ConfChangeType::RemoveNode {
+            if let Some(last_heartbeat) = peer_heartbeat.get(&peer.get_id()) {
+                // peer itself is *not* slow peer, but current slow peer is >= total peers/2
+                if last_heartbeat.elapsed() <= MAX_MISSED_HEARTBEATS * cfg.raft_heartbeat_interval()
+                    && slow_peer_count >= peer_heartbeat.len() as u32 / 2
```

Comment:
maybe use `peer.raft_group.raft.prs().has_quorum(peers_except_slow)`

Reply:
@Connor1996 Good point. I thought it initially too, but end up this simple check instead because of unit test issue.  Do you know how to construct a proper ProgressTracker in unit test (like the test_check_conf_change_upon_slow_peers I added). 
I found it's very hard to create a valid ProgressTracker instance for unit testing. And it's even harder to create a proper peer or raft_group object.

Reply:
OK, used a quite tedious way to create a usable ProgressTracker object.

Reply:
@Connor1996 It turns out that there's a problem with has_quorum() here. Because the remove_node cmd is supposed to change the raft group members, but has_quorum() assumes the members are not going to be changed when calculating quorum.  
So for example, if we have two peers' raft group and want to remove a peer, with this change it will not succeed. Because the remaining one peer will not achieve quorum under current progress object.